python/eos/figure/figure.py

```diff
@@ -0,0 +1,142 @@
+# Copyright (c) 2023 Philip Lueghausen
+#
+# This file is part of the EOS project. EOS is free software;
+# you can redistribute it and/or modify it under the terms of the GNU General
+# Public License version 2, as published by the Free Software Foundation.
+#
+# EOS is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
+# Place, Suite 330, Boston, MA  02111-1307  USA
+
+from .plot import PlotFactory
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+class Figure:
+    def __init__(self, figure, axes, descriptions:list): # plots -> descriptions
+        self.figure = figure
+        self.axes = axes
+        self.plots = []
+
+        if not descriptions:
+            raise ValueError(f"Argument 'descriptions' is empty list")
+
+        for ax, plot_description in zip(axes, descriptions):
+            if plot_description is None:
+                self.plots.append(PlotFactory.make(ax, [], plot_type='empty'))
+                continue
+
+            if 'description' not in plot_description:
+                raise ValueError(f"Plot does not contain element 'description'")
+            if 'items' not in plot_description:
+                raise ValueError(f"Plot does not contain element 'items'")
+
+            contents = plot_description['items']
+            plot_description = PlotFactory.sanitize(plot_description['description'])
+
+            # Implicit default for plot type of Figure
+            if 'type' not in plot_description:
+                plot_description['type'] = '2dplot'
+
+            self.plots.append(PlotFactory.make(ax, contents, **plot_description))
+
+
+    def draw(self):
+        for plot in self.plots:
+            plot.draw()
+
+
+class SingleFigure(Figure):
+    """Represent a figure with a single set of axes.
+
+    Use the value of 'plot' as a set of arguments to construct a Figure.
+
+    :param plot: The plot's data, as a dictionary containing the 'description' and the 'contents'.
+    :type plot: dict
+    """
+
+    def __init__(self, description):
+        figure, axes = plt.subplots()
+        super().__init__(figure, [axes,], [description,])
+
+
+class GridFigure(Figure):
+    """Represent a figure with several axes arranged in a grid.
+
+    :param nrows: The number of rows in the grid.
+    :type nrows: int
+    :param ncols: The number of columns in the grid.
+    :type ncols: int
+    :param plots: The list of plots' data, as a list of dictionaries each containing the 'description' and the 'contents' for a single plot.
+    :type plot: list[dict] or iterable[dict]
+    """
+
+    def __init__(self, nrows, ncols, descriptions, **kwargs):
```

Comment:
Is this a convenient order of parameters? I would probably move ``descriptions`` to before ``nrows, ncols``. But maybe there is a good reason for the current ordering?